cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaMetadataStateManager.java

```diff
@@ -0,0 +1,466 @@
+                if (isKRaftAnnoEnabled()) {
+                    return KafkaMetadataConfigurationState.KRAFT;
+                // rollback
+                } else if (isKRaftAnnoRollback()) {
+                    // ZooKeeper configured and migration enabled on both controllers and brokers
+                    return KafkaMetadataConfigurationState.MIGRATION;
+                } else {
+                    // ZooKeeper still configured on controllers, removed on brokers
+                    return KafkaMetadataConfigurationState.POST_MIGRATION;
+                }
+            }
+            case PreKRaft, KRaft -> {
+                return KafkaMetadataConfigurationState.KRAFT;
+            }
+        }
+        // this should never happen
+        throw new IllegalArgumentException("Invalid internal Kafka metadata state: " + this.metadataState);
+    }
+
+    /**
+     * Check for the status of the Kafka metadata migration
+     *
+     * @param reconciliation    Reconciliation information
+     * @param clusterCaCertSecret   Secret with the Cluster CA public key
+     * @param coKeySecret   Secret with the Cluster CA private key
+     * @param controllerPodName Name of the quorum controller leader pod
+     */
+    public void checkMigrationInProgress(Reconciliation reconciliation, Secret clusterCaCertSecret, Secret coKeySecret, String controllerPodName) {
+        if (this.kafkaAgentClient == null) {
+            this.kafkaAgentClient = new KafkaAgentClient(reconciliation, reconciliation.name(), reconciliation.namespace(), clusterCaCertSecret, coKeySecret);
+        }
+        KRaftMigrationState kraftMigrationState = this.kafkaAgentClient.getKRaftMigrationState(controllerPodName);
+        LOGGER.infoCr(reconciliation, "ZooKeeper to KRaft migration state {} checked on controller {}", kraftMigrationState.state(), controllerPodName);
+        if (kraftMigrationState.state() == -1) {
+            throw new RuntimeException("Failed to get the ZooKeeper to KRaft migration state");
+        }
+        this.isMigrationDone = kraftMigrationState.isMigrationDone();
+    }
+
+    /**
+     * If the KRaft migration process is in the rollback phase, this method deletes the /controller znode from ZooKeeper
+     * to allow the brokers, which are now in ZooKeeper mode again, to elect a new controller among them taking the
+     * KRaft controllers out of the picture
+     *
+     * @param reconciliation    Reconciliation information
+     * @param clusterCaCertSecret   Secret with the Cluster CA public key
+     * @param coKeySecret   Secret with the Cluster CA private key
+     * @param operationTimeoutMs    Timeout to be set on the ZooKeeper request configuration
+     * @param zkConnectionString    Connection string to the ZooKeeper ensemble to connect to
+     */
+    public void maybeDeleteZooKeeperControllerZnode(Reconciliation reconciliation, Secret clusterCaCertSecret, Secret coKeySecret, long operationTimeoutMs, String zkConnectionString) {
```

Comment:
This does not seem to be related to the Migration state and its management. Maybe this should be a static method in some MigrationUtils class?

Reply:
I moved to a new `KRaftMigrationUtils` class.